### PR TITLE
React views

### PR DESCRIPTION
**Pull request by markivancho**

### CODE REVIEW

File: app/controllers/api/v1/timelogs_controller.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      else 
+        @timelogs = current_user.timelogs.date_range(params[:start_time], params[:end_time]).paginate(:page => (params[:page].to_i+1).to_s, :per_page => params[:limit])
+        authorize @timelogs
+      end    
+    else
+      if params[:limit] == "0"
+        @timelogs = current_user.timelogs.count
+      else 
+        @timelogs = current_user.timelogs.paginate(:page => (params[:page].to_i+1).to_s, :per_page => params[:limit])
+        authorize @timelogs
```
Comment (jizak) [inline app/controllers/api/v1/timelogs_controller.rb]: DRY.

File: app/controllers/api/v1/timelogs_controller.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
+require 'trello'
+class Api::V1::TimelogsController < Api::V1::BaseController
+  before_action :authenticate_user!
+  before_action :load_timelog, only: [:update, :destroy]
+  before_action :load_trello_service, only: [:trello_cards]
+
+  def index
```
Comment (jizak) [inline app/controllers/api/v1/timelogs_controller.rb:7]: I am sure this action can be done much better. To many `if else`

File: app/controllers/api/v1/timelogs_controller.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  def trello_cards
+    respond_with @trello_service.cards
+  end
+
+  def create
+    @timelog = current_user.timelogs.build(timelogs_params)
+    authorize @timelog
+    respond_with :api, :v1, @timelog
```
Comment (jizak) [inline app/controllers/api/v1/timelogs_controller.rb]: It makes sense to save a record and after that, we can respond it.
What happens if the record is not saved? Will user see error messages?

File: app/controllers/api/v1/timelogs_controller.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  def create
+    @timelog = current_user.timelogs.build(timelogs_params)
+    authorize @timelog
+    respond_with :api, :v1, @timelog 
+    @timelog.save
+  end
+
+  def update
+    @timelog.update_attributes(timelogs_params)
```
Comment (jizak) [inline app/controllers/api/v1/timelogs_controller.rb]: Again, it is not clear what happens if record is not saved.

File: app/javascript/packs/body.jsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import PropTypes from 'prop-types'
+import ReactPaginate from 'react-paginate';
+import Timelogs from './timelogs.jsx'
+import NewTimelog from './new_timelog.jsx'
+
+class Body extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = { timelogs: [], trello_cards: [], page: 0, pageCount: 1, perPage: 3, start_time: 0, end_time: 0 };
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: You should follow https://github.com/airbnb/javascript rules.

For readability and git maintainability.
```
this.state = {
  some: 'some',
}
```

File: app/javascript/packs/body.jsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+import ReactDOM from 'react-dom'
+import PropTypes from 'prop-types'
+import ReactPaginate from 'react-paginate';
+import Timelogs from './timelogs.jsx'
+import NewTimelog from './new_timelog.jsx'
+
+class Body extends React.Component {
+
+  constructor(props) {
+    super(props);
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: The constructor accepts props and context. You basically removed context from this component.

```
constructor(...args) {
  super(...args);
}
```

Here you are passing all the arguments to parent constructor. See spread operator ES6

File: app/javascript/packs/body.jsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    this.handleSubmit = this.handleSubmit.bind(this, this.state);
+    this.handleDelete = this.handleDelete.bind(this);
+    this.handleUpdate = this.handleUpdate.bind(this);
+    this.removeTimelog = this.removeTimelog.bind(this, this.state);
+    this.updateTimelogs = this.updateTimelogs.bind(this, this.state);
+    this.loadTimelogs = this.loadTimelogs.bind(this, this.state);
+    this.filterByDuration = this.filterByDuration.bind(this, this.state);
+    this.filterByStartTime = this.filterByStartTime.bind(this, this.state);
+    this.filterByEndTime = this.filterByEndTime.bind(this, this.state);
+    this.filterByTimeRange = this.filterByTimeRange.bind(this, this.state);
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: To bind only `this` in most cases is enough. Did you have any problems that you've decided to bind `this.state` as well?

File: app/javascript/packs/body.jsx [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    this.discardFilter = this.discardFilter.bind(this);
+  }
+
+  loadTimelogs() {
+    //Get all timelogs for pageCount
+    $.ajax({
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Maybe take a look into https://github.com/github/fetch? jQuery with React is redundant.

File: app/javascript/packs/body.jsx [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+      type     : 'GET',
+      success: data => {
+        this.setState({pageCount: Math.ceil(data / this.state.perPage)});
+      }
+    });
+    //Get only few timelogs for current page
+    $.ajax({
+      url      : '/api/v1/timelogs',
+      beforeSend: function(xhr) {xhr.setRequestHeader('X-CSRF-Token', $('meta[name="csrf-token"]').attr('content'))},
+      data     : {limit: this.state.perPage, page: this.state.page, start_time: this.state.start_time, end_time: this.state.end_time},
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Again, use this structure for objects with more than one key:

```
{
  some: some,
}
```

File: app/javascript/packs/body.jsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        this.updateTimelogs(timelog);
+      }
+    });
+  }
+
+  updateTimelogs(timelog) {
+    this.loadTimelogs();
+    //var timelogs = this.state.timelogs.filter((i) => { return i.id != timelog.id });
+    //timelogs.push(timelog);
+    //this.setState({timelogs: timelogs })
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Commented code should be deleted

File: app/javascript/packs/body.jsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+      } else {
+        return (
+          <div>
+            Filter by:
+            <button onClick={this.filterByDuration}>Duration</button>
+            <button onClick={this.filterByStartTime}>Start time</button>
+            <button onClick={this.filterByEndTime}>End time</button><br /><br />
+            Select time range:
+            <input type="datetime-local" ref='start_time' onChange={e => this.setState({start_time: e.target.value})}/>
+            <input type="datetime-local" ref='end_time' onChange={e => this.setState({end_time: e.target.value})}/>
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Move it out to the method of class

File: app/javascript/packs/hello_react.jsx [hunk 12]
```diff
@@ -7,7 +7,7 @@ import ReactDOM from 'react-dom'
 import PropTypes from 'prop-types'
 
 const Hello = props => (
-  <div>Hello {props.name}!</div>
+  <div> Hello {props.name}! </div>
```
Comment (infiman) [inline app/javascript/packs/hello_react.jsx]: Is this file still needed?

File: app/javascript/packs/new_timelog.jsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+import PropTypes from 'prop-types'
+import Select from 'react-normalized-select'
+
+class NewTimelog extends React.Component {
+
+  constructor(props) {
```
Comment (infiman) [inline app/javascript/packs/new_timelog.jsx]: Pass `...args` to super from the constructor. Look above for reference why and how to do this.

File: app/javascript/packs/new_timelog.jsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    super(props);
+    this.state = { card: false };
+  }
+
+  handleClick() {
+    var start_time  = this.refs.start_time.value;
```
Comment (infiman) [inline app/javascript/packs/new_timelog.jsx]: Look for `const` and `let` in ES6. Consider in future to use `let` where was `var` and `const` for all immutable variables.

File: app/javascript/packs/new_timelog.jsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    var duration = this.refs.duration.value;
+    var trello_card = this.state.card || this.props.trello_cards[0];
+      $.ajax({
+        url: '/api/v1/timelogs',
+        type: 'POST',
+        beforeSend: function(xhr) {xhr.setRequestHeader('X-CSRF-Token', $('meta[name="csrf-token"]').attr('content'))},
```
Comment (infiman) [inline app/javascript/packs/new_timelog.jsx]: You can do it with pure JavaScript and then you won't need jQuery

File: app/javascript/packs/new_timelog.jsx [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        });
+    }
+  render() {
+    return (
+      <div id="newtimelog">
+        <h3>Create a new timelog!</h3>
+        <input type="datetime-local" ref='start_time' /><br />
+        <input type="number" ref='duration' placeholder='Enter the duration in minutes' /><br />
+        <Select className="mySelect" onChange={e => this.setState({card: e.target.value})}>
```
Comment (infiman) [inline app/javascript/packs/new_timelog.jsx]: Move outside to the class method.

File: app/javascript/packs/new_timelog.jsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+          {this.props.trello_cards.map(option => <option key={option} value={option}>{option}</option>)}
+          }
+        </Select>
+        <br />
+        <button onClick={this.handleClick.bind(this)}>Create</button><br />
```
Comment (infiman) [inline app/javascript/packs/new_timelog.jsx]: Bind all the needed functions in the constructor. It's a performance issue. Every time when React calling `render()` method, `.bind(...)` will be called, and its memory and process consuming issue. If you will bind in the constructor, then you will avoid it since constructor will be called only once.

File: app/javascript/packs/timelog.jsx [hunk 18]
```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+import PropTypes from 'prop-types'
+import Select from 'react-normalized-select'
+
+class Timelog extends React.Component {
+
+  constructor(props) {
+    super(props);
```
Comment (infiman) [inline app/javascript/packs/timelog.jsx]: Same context issue

File: app/javascript/packs/timelog.jsx [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+      var duration = this.refs.duration.value;
+      var trello_card = this.state.card;
+      var timelog = {id: id , start_time: start_time , duration: duration, trello_card: trello_card};
+      this.props.handleUpdate(timelog);
+    }
+    this.setState({ editable: !this.state.editable })
+  }
+
+  render() {
+    var start_time = this.state.editable ? <input type='datetime-local' ref='start_time' defaultValue={this.props.timelog.start_time.substring(0, this.props.timelog.start_time.length-5)}/>: <p>Start time: {this.props.timelog.start_time.substring(0, this.props.timelog.start_time.length-5)}</p>;
```
Comment (infiman) [inline app/javascript/packs/timelog.jsx]: It's a mess. Consider moving to separate getter methods. And use formatting for ternary operator

File: app/javascript/packs/timelogs.jsx [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  onUpdate(timelog) {
+    this.props.onUpdate(timelog);
+  }
+
+  render() {
+    var timelogs = this.props.timelogs.map((timelog) => {
+      return (
+        <Timelog id={timelog.id} key={timelog.id.toString()} trello_cards={this.props.trello_cards} timelog={timelog} handleDelete={this.handleDelete.bind(this, timelog.id)} handleUpdate={this.onUpdate.bind(this)} />
```
Comment (infiman) [inline app/javascript/packs/timelogs.jsx]: Don't bind in the render method.

File: app/javascript/packs/body.jsx [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  updateTimelogs(timelog) {
+    this.loadTimelogs();
+    //var timelogs = this.state.timelogs.filter((i) => { return i.id != timelog.id });
+    //timelogs.push(timelog);
+    //this.setState({timelogs: timelogs })
+  }
+
+  handlePageClick = (data) => {
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Bind in the constructor and rewrite as a method of the class.

File: app/javascript/packs/body.jsx [hunk 22]
```diff
@@ -27,7 +27,7 @@ class Body extends React.Component {
     //Get all timelogs for pageCount
     $.ajax({
       url      : '/api/v1/timelogs',
-      beforeSend: function(xhr) {xhr.setRequestHeader('X-CSRF-Token', $('meta[name="csrf-token"]').attr('content'))},
+      beforeSend: function(xhr) {xhr.setRequestHeader('X-CSRF-Token', $('meta[name="csrf-token"]').attr('content'));},
```
Comment (infiman) [inline app/javascript/packs/body.jsx]: Also, it's nicer to use arrow function here

**Review by jizak — APPROVED** [review]
